File: api/razorpay-webhook.js
import { createHmac } from 'crypto';
import { readFileSync, writeFileSync, existsSync } from 'fs';
import { join } from 'path';

const EVENTS_FILE = join('/tmp', 'razorpay-webhook-events.json');

function loadEvents() {
  try {
    if (existsSync(EVENTS_FILE)) {
      return JSON.parse(readFileSync(EVENTS_FILE, 'utf-8'));
    }
  } catch {}
  return [];
}

function saveEvents(events) {
  // Keep only the latest 200 events
  const trimmed = events.slice(0, 200);
  writeFileSync(EVENTS_FILE, JSON.stringify(trimmed), 'utf-8');
}

function verifySignature(body, signature, secret) {
  const expectedSignature = createHmac('sha256', secret)
    .update(body)
    .digest('hex');
  return expectedSignature === signature;
}

export default async function handler(req, res) {
  // GET — return stored webhook events
  if (req.method === 'GET') {
    const events = loadEvents();
    return res.status(200).json(events);
  }

  // POST — receive and process webhook
  if (req.method === 'POST') {
    const secret = process.env.RAZORPAY_WEBHOOK_SECRET;

    // Read raw body for signature verification
    const rawBody = typeof req.body === 'string' ? req.body : JSON.stringify(req.body);

    // Verify signature if secret is configured
    if (secret) {
      const signature = req.headers['x-razorpay-signature'];
      if (!signature) {
        return res.status(401).json({ error: 'Missing X-Razorpay-Signature header' });
      }
      if (!verifySignature(rawBody, signature, secret)) {
        return res.status(401).json({ error: 'Invalid webhook signature' });
      }
    }

    const data = typeof req.body === 'string' ? JSON.parse(req.body) : req.body;
    const event = data.event;

    const logEntry = {
      receivedAt: new Date().toISOString(),
      event: event || 'unknown',
      processed: false,
    };

    if (event === 'payment.captured') {
      const payment = data.payload && data.payload.payment && data.payload.payment.entity;
      if (payment) {
        const amountRupees = (payment.amount || 0) / 100;
        let name = '';
        try {
          if (payment.notes && typeof payment.notes === 'object') {
            name = payment.notes.name || payment.notes.registered_name || '';
          }
        } catch {}
        if (!name && payment.card && payment.card.name) name = payment.card.name;
        if (!name) name = (payment.email || '').split('@')[0];

        logEntry.paymentId = payment.id;
        logEntry.email = (payment.email || '').toLowerCase().trim();
        logEntry.name = name.trim();
        logEntry.phone = (payment.contact || '').trim();
        logEntry.amount = amountRupees;
        logEntry.currency = payment.currency || 'INR';
        logEntry.createdAt = payment.created_at;
        logEntry.processed = true;
      }
    }

    // Append to stored events
    const events = loadEvents();
    events.unshift(logEntry);
    saveEvents(events);

    return res.status(200).json({ status: 'ok', event: logEntry.event, processed: logEntry.processed });
  }

  return res.status(405).json({ error: 'Method not allowed' });
}

File: api/razorpay_webhook.py
"""Razorpay webhook receiver (Vercel serverless function)."""
import json
import os
import time
import hmac
import hashlib
from http.server import BaseHTTPRequestHandler


RECENT_EVENTS = []
MAX_RECENT_EVENTS = 20


def _get_secret():
    return (
        os.getenv("RAZORPAY_WEBHOOK_SECRET")
        or os.getenv("BOOKLEAF_RAZORPAY_WEBHOOK_SECRET")
        or ""
    )


def _append_recent(event):
    RECENT_EVENTS.append(event)
    if len(RECENT_EVENTS) > MAX_RECENT_EVENTS:
        del RECENT_EVENTS[:-MAX_RECENT_EVENTS]


class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        secret = _get_secret()
        self._send_json(
            200,
            {
                "ok": True,
                "configured": bool(secret),
                "recent": list(reversed(RECENT_EVENTS[-10:])),
            },
        )

    def do_POST(self):
        secret = _get_secret()
        if not secret:
            self._send_json(500, {"error": "RAZORPAY_WEBHOOK_SECRET not configured"})
            return

        content_length = int(self.headers.get("Content-Length", "0") or "0")
        raw_body = self.rfile.read(content_length) if content_length else b""

        signature = (self.headers.get("X-Razorpay-Signature") or "").strip()
        if not signature:
            self._send_json(400, {"error": "Missing X-Razorpay-Signature"})
            return

        expected = hmac.new(secret.encode("utf-8"), raw_body, hashlib.sha256).hexdigest()
        if not hmac.compare_digest(expected, signature):
            self._send_json(401, {"error": "Invalid webhook signature"})
            return

        event_name = "unknown"
        payload = None
        if raw_body:
            try:
                payload = json.loads(raw_body.decode("utf-8"))
                event_name = payload.get("event") or "unknown"
            except Exception:
                # Keep accepting the webhook even if body parsing fails; signature already passed.
                payload = None

        _append_recent(
            {
                "receivedAt": int(time.time()),
                "event": event_name,
                "contentLength": len(raw_body),
            }
        )

        # TODO: Persist/process events if/when business logic is finalized.
        self._send_json(200, {"ok": True, "event": event_name})

    def do_OPTIONS(self):
        self.send_response(200)
        self._cors()
        self.end_headers()

    def _send_json(self, status, body):
        data = json.dumps(body).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Cache-Control", "no-store")
        self._cors()
        self.end_headers()
        self.wfile.write(data)

    def _cors(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type, X-Razorpay-Signature")

